File: src/test/java/kitchenpos/application/MenuServiceTest.java
package kitchenpos.application;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import kitchenpos.domain.Menu;
import kitchenpos.domain.MenuGroup;
import kitchenpos.domain.MenuGroupRepository;
import kitchenpos.domain.MenuProduct;
import kitchenpos.domain.MenuRepository;
import kitchenpos.domain.Product;
import kitchenpos.domain.ProductRepository;
import kitchenpos.infra.PurgomalumClient;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class MenuServiceTest {

    private final MenuRepository menuRepository = mock(MenuRepository.class);
    private final MenuGroupRepository menuGroupRepository = mock(MenuGroupRepository.class);
    private final ProductRepository productRepository = mock(ProductRepository.class);
    private final PurgomalumClient purgomalumClient = mock(PurgomalumClient.class);

    private final MenuService menuService = new MenuService(menuRepository, menuGroupRepository,
        productRepository, purgomalumClient);

    @ParameterizedTest
    @DisplayName("메뉴 등록 시 가격이 없거나 0 보다 작으면 예외를 발생 시킨다.")
    @ValueSource(ints = {-1, -2})
    void create(int price) {
        var request = new Menu();
        request.setPrice(BigDecimal.valueOf(price));

        assertThatThrownBy(
            () -> menuService.create(request)
        ).isInstanceOf(IllegalArgumentException.class)
            .hasMessage("메뉴 가격은 0보다 작을 수 없습니다.");
    }

    @Test
    @DisplayName("메뉴 등록 시 요청 상품이 없는 경우 예외를 발생 시킨다.")
    void createWithoutProduct() {
        var request = new Menu();
        request.setMenuGroupId(UUID.randomUUID());
        request.setPrice(BigDecimal.valueOf(10_000));
        when(menuGroupRepository.findById(any())).thenReturn(Optional.of(new MenuGroup()));

        assertThatThrownBy(
            () -> menuService.create(request)
        ).isInstanceOf(IllegalArgumentException.class)
            .hasMessage("메뉴 구성 상품이 없습니다.");
    }

    @Test
    @DisplayName("메뉴 등록 시 요청 상품 수량이 0보다 작은 경우 예외를 발생 시킨다.")
    void createWithEmptyProduct() {
        var request = new Menu();
        request.setMenuGroupId(UUID.randomUUID());
        request.setPrice(BigDecimal.valueOf(10_000));

        var menuProduct = new MenuProduct();
        menuProduct.setQuantity(-1L);
        request.setMenuProducts(List.of(menuProduct));

        when(menuGroupRepository.findById(any())).thenReturn(Optional.of(new MenuGroup()));
        when(productRepository.findAllByIdIn(any())).thenReturn(List.of(new Product()));

        assertThatThrownBy(
            () -> menuService.create(request)
        ).isInstanceOf(IllegalArgumentException.class)
            .hasMessage("메뉴 구성 상품 수량은 0보다 작을 수 없습니다.");
    }

    @Test
    @DisplayName("메뉴 가격이 구성 상품 가격 합 보다 클 경우 예외를 발생 시킨다.")
    void createWithWrongPrice() {
        var menuProduct = new MenuProduct();
        menuProduct.setQuantity(2L);

        var product = new Product();
        product.setPrice(BigDecimal.valueOf(10_000));

        var request = new Menu();
        request.setMenuGroupId(UUID.randomUUID());
        request.setPrice(BigDecimal.valueOf(50_000));
        request.setMenuProducts(List.of(menuProduct));

        when(menuGroupRepository.findById(any())).thenReturn(Optional.of(new MenuGroup()));
        when(productRepository.findAllByIdIn(any())).thenReturn(List.of(product));
        when(productRepository.findById(any())).thenReturn(Optional.of(product));

        assertThatThrownBy(
            () -> menuService.create(request)
        ).isInstanceOf(IllegalArgumentException.class)
            .hasMessage("메뉴 가격은 상품의 총 합보다 클 수 없습니다.");
    }

    @Test
    @DisplayName("메뉴 이름에 비속어가 포함되면 예외를 발생 시킨다.")
    void createWithBadName() {
        var menuProduct = new MenuProduct();
        menuProduct.setQuantity(2L);

        var product = new Product();
        product.setPrice(BigDecimal.valueOf(10_000));

        var request = new Menu();
        request.setMenuGroupId(UUID.randomUUID());
        request.setPrice(BigDecimal.valueOf(10_000));
        request.setMenuProducts(List.of(menuProduct));
        request.setName("욕설");

        when(menuGroupRepository.findById(any())).thenReturn(Optional.of(new MenuGroup()));
        when(productRepository.findAllByIdIn(any())).thenReturn(List.of(product));
        when(productRepository.findById(any())).thenReturn(Optional.of(product));
        when(purgomalumClient.containsProfanity("욕설")).thenReturn(Boolean.TRUE);

        assertThatThrownBy(
            () -> menuService.create(request)
        ).isInstanceOf(IllegalArgumentException.class)
            .hasMessage("비속어는 사용할 수 없습니다.");
    }


    @Test
    @DisplayName("가격 수정 시 0보다 작은 값을 입력하면 예외를 발생 시킨다.")
    void changePrice() {
        var request = new Menu();
        request.setPrice(BigDecimal.valueOf(-1));

        assertThatThrownBy(
            () -> menuService.changePrice(UUID.randomUUID(), request)
        ).isInstanceOf(IllegalArgumentException.class)
            .hasMessage("메뉴 가격은 0보다 작을 수 없습니다.");
    }

    @Test
    @DisplayName("가격 수정 시 메뉴 가격이 구성 상품의 총합보다 클 경우 예외를 발생 시킨다.")
    void changePriceWithGreaterThanProductPrice() {
        var request = new Menu();
        request.setPrice(BigDecimal.valueOf(50_000));

        var menu = getMenu();
        when(menuRepository.findById(any())).thenReturn(Optional.of(menu));

        assertThatThrownBy(
            () -> menuService.changePrice(UUID.randomUUID(), request)
        ).isInstanceOf(IllegalArgumentException.class)
            .hasMessage("메뉴 가격은 상품의 총 합보다 클 수 없습니다.");
    }

    @NotNull
    private static Menu getMenu() {
        var product = new Product();
        product.setPrice(BigDecimal.valueOf(10_000));

        var menuProduct = new MenuProduct();
        menuProduct.setProduct(product);
        menuProduct.setQuantity(1);

        var menu = new Menu();
        menu.setMenuProducts(List.of(menuProduct));
        return menu;
    }

    @Test
    @DisplayName("메뉴 표시하기 시 메뉴 가격이 구성 상품의 총합보다 클 경우 예외를 발생 시킨다.")
    void displayWithBadPrice() {
        var menu = getMenu();
        menu.setPrice(BigDecimal.valueOf(50_000));

        when(menuRepository.findById(any())).thenReturn(Optional.of(menu));

        assertThatThrownBy(
            () -> menuService.display(UUID.randomUUID())
        ).isInstanceOf(IllegalStateException.class)
            .hasMessage("메뉴 가격은 상품의 총 합보다 클 수 없습니다.");
    }

    @Test
    @DisplayName("메뉴 표시하기")
    void display() {
        var menu = getMenu();
        menu.setPrice(BigDecimal.valueOf(10_000));

        when(menuRepository.findById(any())).thenReturn(Optional.of(menu));

        var response = menuService.display(UUID.randomUUID());
        assertThat(response.isDisplayed()).isTrue();
    }

    @Test
    @DisplayName("메뉴 숨기기")
    void hide() {
        var menu = getMenu();
        menu.setPrice(BigDecimal.valueOf(10_000));

        when(menuRepository.findById(any())).thenReturn(Optional.of(menu));

        var response = menuService.hide(UUID.randomUUID());
        assertThat(response.isDisplayed()).isFalse();
    }
}

File: src/test/java/kitchenpos/application/MenuServiceTest.kt
package kitchenpos.application

import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import kitchenpos.domain.MenuGroupRepository
import kitchenpos.domain.MenuProduct
import kitchenpos.domain.MenuRepository
import kitchenpos.domain.ProductRepository
import kitchenpos.infra.PurgomalumClient
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.NullSource
import org.junit.jupiter.params.provider.ValueSource
import random.randomMenu
import random.randomMenuId
import random.randomMenuProduct
import random.randomProduct
import java.math.BigDecimal
import java.util.*

class MenuServiceTest {
    private val menuRepository: MenuRepository = mockk()
    private val menuGroupRepository: MenuGroupRepository = mockk()
    private val productRepository: ProductRepository = mockk()
    private val purgomalumClient: PurgomalumClient = mockk()
    private lateinit var menuService: MenuService

    @BeforeEach
    fun setUp() {
        menuService = MenuService(menuRepository, menuGroupRepository, productRepository, purgomalumClient)
    }

    @Nested
    inner class `메뉴 생성시` {

        @ParameterizedTest
        @NullSource
        @ValueSource(longs = [-1])
        fun `가격은 필수값이며, 0보다 커야한다 아닌 경우 예외 발생`(input: Long?) {
            // given
            val menu = randomMenu(price = input?.let { BigDecimal.valueOf(it) })
            // when then
            assertThrows<IllegalArgumentException> {
                menuService.create(menu)
            }
        }

        @Test
        fun `생성하려는 하는 메뉴의 그룹이 존재해야한다 아닌 경우 예외 발생`() {
            // given
            val menu = randomMenu()
            every { menuGroupRepository.findById(menu.menuGroupId) } returns Optional.empty()

            // when then
            assertThrows<NoSuchElementException> {
                menuService.create(menu)
            }
        }

        @Test
        fun `메뉴 상품은 필수값이며, 최소 1개 이상이어야 한다 아닌 경우 예외 발생`() {
            // given
            val menu = randomMenu(menuProducts = emptyList())
            every { menuGroupRepository.findById(menu.menuGroupId) } returns Optional.of(menu.menuGroup)

            // when then
            assertThrows<IllegalArgumentException> {
                menuService.create(menu)
            }
        }

        @Test
        fun `메뉴 상품은 반드시 상품 테이블에 존재해야 한다 아닌 경우 에러 발생`() {
            // given
            val menuProducts = listOf(randomMenuProduct(), randomMenuProduct())
            val menu = randomMenu(menuProducts = menuProducts)
            every { menuGroupRepository.findById(menu.menuGroupId) } returns Optional.of(menu.menuGroup)
            every { productRepository.findAllByIdIn(menuProducts.map(MenuProduct::getProductId)) } returns emptyList()

            // when then
            assertThrows<IllegalArgumentException> {
                menuService.create(menu)
            }
        }

        @ParameterizedTest
        @ValueSource(longs = [-1, -2])
        fun `메뉴 상품의 수량은 0보다 커야한다 아닌 경우 예외 발생`(input: Long) {
            // given
            val menuProducts = listOf(randomMenuProduct(quantity = input), randomMenuProduct(quantity = input))
            val menu = randomMenu(menuProducts = menuProducts)
            every { menuGroupRepository.findById(menu.menuGroupId) } returns Optional.of(menu.menuGroup)
            every { productRepository.findAllByIdIn(menuProducts.map(MenuProduct::getProductId)) } returns listOf(
                randomProduct(),
                randomProduct()
            )

            // when then
            assertThrows<IllegalArgumentException> {
                menuService.create(menu)
            }
        }

        @Test
        fun `메뉴 가격이 메뉴 상품들의 합산 가격보다 크지 않아야 한다 아닌 경우 에러 발생`() {
            // given
            val product = randomProduct(price = BigDecimal.valueOf(100))
            val menuProducts = listOf(randomMenuProduct(quantity = 1, product = product))
            val menu = randomMenu(menuProducts = menuProducts, price = BigDecimal.valueOf(250))
            every { menuGroupRepository.findById(menu.menuGroupId) } returns Optional.of(menu.menuGroup)
            every { productRepository.findAllByIdIn(menuProducts.map(MenuProduct::getProductId)) } returns listOf(
                randomProduct(),
                randomProduct()
            )
            every { productRepository.findById(any()) } returns Optional.of(product)

            // when then
            assertThrows<IllegalArgumentException> {
                menuService.create(menu)
            }
        }

        @Test
        fun `메뉴명은 필수값이다 아닐 경우 에러 발생`() {
            // given
            val product = randomProduct(price = BigDecimal.valueOf(100))
            val menuProducts = listOf(randomMenuProduct(quantity = 1, product = product))
            val menu = randomMenu(menuProducts = menuProducts, price = BigDecimal.valueOf(80), name = null)
            every { menuGroupRepository.findById(menu.menuGroupId) } returns Optional.of(menu.menuGroup)
            every { productRepository.findAllByIdIn(menuProducts.map(MenuProduct::getProductId)) } returns listOf(
                randomProduct(),
                randomProduct()
            )
            every { productRepository.findById(any()) } returns Optional.of(product)

            // when then
            assertThrows<IllegalArgumentException> {
                menuService.create(menu)
            }
        }

        @Test
        fun `메뉴명은 비속어가 포함되지 않아야 한다 아닐 경우 에러 발생`() {
            // given
            val product = randomProduct(price = BigDecimal.valueOf(100))
            val menuProducts = listOf(randomMenuProduct(quantity = 1, product = product))
            val menu = randomMenu(menuProducts = menuProducts, price = BigDecimal.valueOf(80))
            every { menuGroupRepository.findById(menu.menuGroupId) } returns Optional.of(menu.menuGroup)
            every { productRepository.findAllByIdIn(menuProducts.map(MenuProduct::getProductId)) } returns listOf(
                randomProduct(),
                randomProduct()
            )
            every { productRepository.findById(any()) } returns Optional.of(product)
            every { purgomalumClient.containsProfanity(menu.name) } returns true

            // when then
            assertThrows<IllegalArgumentException> {
                menuService.create(menu)
            }
        }
    }

    @Nested
    inner class `메뉴 수정시` {

        @ParameterizedTest
        @NullSource
        @ValueSource(longs = [-1])
        fun `가격은 필수값이며, 0보다 커야한다 아닌 경우 예외 발생`(input: Long?) {
            // given
            val menu = randomMenu(price = input?.let { BigDecimal.valueOf(it) })
            // when then
            assertThrows<IllegalArgumentException> {
                menuService.changePrice(randomMenuId(), menu)
            }
        }

        @Test
        fun `수정하려는 메뉴가 존재해야 한다 아닌 경우 예외 발생`() {
            // given
            val menu = randomMenu()
            every { menuRepository.findById(menu.id) } returns Optional.empty()

            // when then
            assertThrows<NoSuchElementException> {
                menuService.changePrice(menu.id, menu)
            }
        }

        @Test
        fun `수정된 가격이 메뉴 상품들의 합산 가격보다 크지 않아야 한다 아닌 경우 예외 발생`() {
            // given
            val menu = randomMenu(price = BigDecimal.valueOf(1000))
            every { menuRepository.findById(menu.id) } returns Optional.of(
                randomMenu(
                    menuProducts = listOf(
                        randomMenuProduct(product = randomProduct(price = BigDecimal.valueOf(100)), quantity = 1)
                    )
                )
            )

            // when then
            assertThrows<IllegalArgumentException> {
                menuService.changePrice(menu.id, menu)
            }
        }

        @Test
        fun `가격 정상 변경시, 변경된 가격 반환한다`() {
            // given
            val expectedPrice = BigDecimal.valueOf(100)
            val menu = randomMenu(price = expectedPrice)
            every { menuRepository.findById(menu.id) } returns Optional.of(
                randomMenu(
                    menuProducts = listOf(
                        randomMenuProduct(product = randomProduct(price = BigDecimal.valueOf(100)), quantity = 1)
                    )
                )
            )
            // when
            val result = menuService.changePrice(menu.id, menu)
            // then
            result.price shouldBe expectedPrice
        }
    }

    @Nested
    inner class `메뉴 노출` {

        @Test
        fun `노출하려는 메뉴가 존재해야 한다 아닌 경우 예외 발생`() {
            // given
            val menuId = randomMenuId()
            every { menuRepository.findById(menuId) } returns Optional.empty()

            // when then
            assertThrows<NoSuchElementException> {
                menuService.display(menuId)
            }
        }

        @Test
        fun `메뉴 가격이 메뉴 상품들의 합산 가격보다 크지 않아야 한다 아닌 경우 예외 발생`() {
            val menuId = randomMenuId()
            every { menuRepository.findById(menuId) } returns Optional.of(
                randomMenu(
                    id = menuId,
                    price = BigDecimal.valueOf(1000),
                    menuProducts = listOf(
                        randomMenuProduct(product = randomProduct(price = BigDecimal.valueOf(100)), quantity = 1)
                    )
                )
            )

            // when then
            assertThrows<IllegalStateException> {
                menuService.display(menuId)
            }
        }

        @Test
        fun `정상 노출 반환`() {
            // given
            val menuId = randomMenuId()
            every { menuRepository.findById(menuId) } returns Optional.of(randomMenu(isDisplayed = false, price = BigDecimal.ZERO))

            // when
            val result = menuService.display(menuId)

            // then
            result.isDisplayed shouldBe true
        }
    }

    @Nested
    inner class `메뉴 비노출` {

        @Test
        fun `비노출하려는 메뉴가 존재해야 한다 아닌 경우 예외 발생`() {
            // given
            val menuId = randomMenuId()
            every { menuRepository.findById(menuId) } returns Optional.empty()

            // when then
            assertThrows<NoSuchElementException> {
                menuService.hide(menuId)
            }
        }

        @Test
        fun `정상 비노출 반환`() {
            // given
            val menuId = randomMenuId()
            every { menuRepository.findById(menuId) } returns Optional.of(randomMenu(isDisplayed = true))

            // when
            val result = menuService.hide(menuId)

            // then
            result.isDisplayed shouldBe false
        }
    }

    @Test
    fun `메뉴 조회`() {
        val expectedMenus = listOf(randomMenu(), randomMenu())
        every { menuRepository.findAll() } returns expectedMenus

        menuService.findAll() shouldBe expectedMenus
    }
}
